accessible/tree.service.js

```diff
@@ -0,0 +1,325 @@
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an 'AS IS' BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Angular2 Service that handles all tree keyboard navigation.
+ * @author madeeha@google.com (Madeeha Ghori)
+ */
+blocklyApp.TreeService = ng.core
+  .Class({
+    constructor: function() {
+      console.log('making a new tree service');
+      this.activeDesc_ = {};
+      this.previousKey_;
+      this.trees = document.getElementsByClassName('blocklyTree');
+    },
+    createId: function(obj) {
+      if (obj && obj.id) {
+        return obj.id;
+      }
+      return 'blockly-' + Blockly.genUid();
+    },
+    setActiveAttribute: function(tree) {
+      if (!document.getElementById(tree.getAttribute(
+          'aria-activedescendant'))) {
+        blocklyApp.debug && console.log('setting tree active descendant');
+        tree.setAttribute('aria-activedescendant', tree.id + '-node0');
+      }
+    },
+    setActiveDesc: function(node, id) {
+      blocklyApp.debug && console.log('setting active descendant for tree ' + id);
+      this.activeDesc_[id] = node;
+    },
+    getActiveDesc: function(id) {
+      return this.activeDesc_[id];
+    },
+    updateSelectedNode: function(node, tree, keepFocus) {
+      blocklyApp.debug && console.log('updating node: ' + node.id);
+      var treeId = tree.id;
+      var activeDesc = this.getActiveDesc(treeId);
+      if (activeDesc) {
+        activeDesc.classList.remove('blocklyActiveDescendant');
+      } else {
+        blocklyApp.debug && console.log('updateSelectedNode: there is no active descendant');
+      }
+      node.classList.add('blocklyActiveDescendant');
+      tree.setAttribute('aria-activedescendant', node.id);
+      this.setActiveDesc(node, treeId);
+      node.setAttribute('aria-selected', 'true');
+
+      // make sure keyboard focus is on tree as a whole
+      // in case before the user was editing a block and keyboard
+      // focus got shifted.
+      if (keepFocus) {
+        tree.focus();
+      }
+    },
+    workspaceButtonKeyHandler: function(e, treeId) {
+      blocklyApp.debug && console.log(e.keyCode);
+      blocklyApp.debug && console.log('inside TreeService');
+      switch (e.keyCode) {
+        case 9:
+          // 16,9: shift, tab
+          if (e.shiftKey) {
+            console.log('shifttabbing');
+            // if the previous key is shift, we're shift-tabbing mode
+            this.goToPreviousTree(treeId, e);
+          } else {
+            // if previous key isn't shift, we're tabbing
+            this.goToNextTree(treeId, e);
+          }
+          break;
+      }
+      this.previousKey_ = e.keyCode;
+    },
+    goToNextTree: function(treeId, e) {
```

Comment:
The way this is written is a bit confusing, given the way 'next' is used. How about:

      for (var i = 0; i < this.trees.length; i++) {
        if (this.trees[i].id == treeId) {
          if (i + 1 < this.trees.length) {
            this.trees[i + 1].focus();
          }
          break;
        }
      }
      e.preventDefault();
      e.stopPropagation();